packages/data-context/src/sources/ProjectDataSource.ts

```diff
@@ -210,6 +211,57 @@ export class ProjectDataSource {
     this._specWatcher.on('unlink', onSpecsChanged)
   }
 
+  private formatGlob (s: string, fallback: string) {
+    let [cleanedVal] = s.replace('{', '').replace('}', '').split(',')
+
+    if (cleanedVal?.includes('*')) {
+      cleanedVal = cleanedVal.replace(/\*/g, fallback)
+    }
+
+    return cleanedVal
+  }
+
+  async defaultSpecFileName () {
+    const defaultFileName = 'cypress/e2e/filename.cy.js'
+
+    try {
+      if (!this.ctx.currentProject || !this.ctx.coreData.currentTestingType) {
+        return null
+      }
+
+      let specPatternSet: string | undefined
+      const { specPattern = [] } = await this.ctx.project.specPatternsForTestingType(this.ctx.currentProject, this.ctx.coreData.currentTestingType)
+
+      if (Array.isArray(specPattern)) {
```

Comment:
This method is so complicated, without unit tests I would be scared to touch this. I think this is a good candidate for some unit tests. There is a lot of conditionals, I feel like we cannot possibly be capturing all the branches in the one e2e test you added.

What if we extract a pure, stateless method? Something like:

```js
// not inside `ProjectDataSource` class directly, maybe declared outside?

export async function getDefaultSpecFileName (currentProject: string, currentTestingType: TestingType) : Promise<string> {
  let specPatternSet: string | undefined
  // ... all the things ...
  return splittedDirname + fileName + extname + ext
}
```

Unit test this, than wrap it in `ProjectDataSource#defaultSpecFileName`.

```js
async defaultSpecFileName () {
  const defaultFileName = 'cypress/e2e/filename.cy.js'

  if (!this.ctx.currentProject || !this.ctx.coreData.currentTestingType) {
    return null
  }

  try {
    return getDefaultSpecFileName(this.ctx.currentProject, this.ctx.coreData.currentTestingType)
  } catch (e) {
    return defaultFileName
  }
}
```

What do you think? We definitely lack a lot of coverage in `sources`, this seems like a good place to start adding some.